lib/service/base.rb

```diff
@@ -257,6 +261,43 @@ def run_step
       end
     end
 
+    # @!visibility private
+    class LockStep < Step
+      include StepsHelpers
+
+      attr_reader :steps
+
+      def initialize(*keys, &block)
+        @keys = keys
+        @steps = []
+        instance_exec(&block)
+      end
+
+      def run_step
+        success =
+          begin
+            DistributedMutex.synchronize(lock_name) do
+              steps.each { |step| step.call(instance, context) }
+              :success
+            end
+          rescue Discourse::ReadOnly
+            :read_only
```

Comment:
Only catching the `Discourse::ReadOnly` exception since they tend to be heavyweight and I'd rather use them for "exceptional" circumstances